File: src/algorithm/java/leetcode/easy/ReversString.java
package algorithm.java.leetcode.easy;

public class ReversString {
    /*
        char[] s = {'h','e','l','l','o'};
        ReversString rs = new ReversString();
        rs.reverseString(s);
     */

    /*
    Example 1:
    Input: s = ["h","e","l","l","o"]
    Output: ["o","l","l","e","h"]
    Example 2:
    Input: s = ["H","a","n","n","a","h"]
    Output: ["h","a","n","n","a","H"]

    Constraints: 1 <= s.length <= 105
    s[i] is a printable ascii character.
     */
    public void reverseString(char[] s) {
        int i = s.length - 1;
        int j = 0;
        /*
        in place algorithm 참고 (https://en.wikipedia.org/wiki/In-place_algorithm)
        array 를 두 개 만들어서 동시에 사용하면 O(n)의 추가 공간을 필요로 하게된다.
        그래서 in place 알고리즘처럼 하나의 어레이로 해결하는 것이 O(1) 을 사용하는 빠른 방법이다.
         */

        while (i >= s.length / 2){
            char c = s[j];
            s[j] = s[i];
            s[i] = c;
            i--;
            j++;
        }
    }
}

File: src/algorithm/kotlin/leetcode/easy/ReversString.kt
package algorithm.kotlin.leetcode.easy

class ReversString {

    /*
    val chars = charArrayOf('h','e','l','l','o')
//    val chars = charArrayOf('H','a','n','n','a','h')
    val rs = ReversString()
    rs.reversStrung(chars)

    You must do this by modifying the input array in-place with O(1) extra memory.
    Example 1:
    Input: s = ["h","e","l","l","o"]
    Output: ["o","l","l","e","h"]
    Example 2:
    Input: s = ["H","a","n","n","a","h"]
    Output: ["h","a","n","n","a","H"]

    Constraints: 1 <= s.length <= 105
    s[i] is a printable ascii character.
 */
    fun reversStrung(s : CharArray) {
        var index = s.size -1
        val size = if (s.size % 2 == 0) s.size / 2  else (s.size -1) / 2
        for (i: Int in 0..< size){
            val x = s[i]
            s[i] = s[index]
            s[index] = x
            index--
        }
        print(s)
    }

    class BookAnswer{
        fun reversStrung(s : CharArray) {
            var start = 0
            var end = s.size -1
            while (start < end){
                //코틀린에서 지원하는 스코프 함수를 사용하면 보다 코틀린스럽게 해결 할 수 있다.
                s[start] = s[end].also { s[end] = s[start] }
                start++
                end--
            }
            print(s)
        }
    }
}